[RISCV][LLVM] Enable atomics for 'Zalrsc'

slachowsky opened the pull request:
The 'A' atomics extension is composed of two subextensions, 'Zaamo' which has
atomic memory operation instructions, and 'Zalrsc' which has load-reserve /
store-conditional instructions.
    
For machines where 'Zalrsc' is present, but 'Zaamo' is not, implement and enable
atomics memory operations through pseudo expansion.  Updates the predication
and lowering control to be more precise about which 'Zaamo'/'Zalrsc' feature
was truly requisite.
    
There will be no functional change to subtargets supporting 'A', while allowing
'Zalrsc' only subtargets to utilize atomics at an increased code footprint.

slachowsky committed:
[RISCV][LLVM] Enable atomics for 'Zalrsc'

The 'A' atomics extension is composed of two subextensions, 'Zaamo' which has
atomic memory operation instructions, and 'Zalrsc' which has load-reserve /
store-conditional instructions.

For machines where 'Zalrsc' is present, but 'Zaamo' is not, implement and enable
atomics memory operations through pseudo expansion.  Updates the predication
and lowering control to be more precise about which 'Zaamo'/'Zalrsc' feature
was truly requisite.

There will be no functional change to subtargets supporting 'A', while allowing
'Zalrsc' only subtargets to utilize atomics at an increased code footprint.

slachowsky committed:
[RISCV][clang] Enable atomics for 'Zalrsc', not just 'A'

'Zalrsc' only subtargets are atomics capable, ensure the RISCV target exposes
this correctly.

slachowsky commented:
@topperc kept me awake rebasing passed 3bec46ff95888c537dabbf19895d48e3f0eb7b35

topperc commented on llvm/lib/Target/RISCV/RISCVExpandAtomicPseudoInsts.cpp:
```diff
@@ -285,6 +378,34 @@ static void doAtomicBinOpExpansion(const RISCVInstrInfo *TII, MachineInstr &MI,
+    break;
+  case AtomicRMWInst::UMin:
+    if (Zext)
+      BuildMI(LoopMBB, DL, TII->get(RISCV::ADD_UW), ScratchReg)
```
ADD_UW belong to Zba not Zbb which I don't think you checked for.

add.uw, min(u), max(u), aren't allowed to be in an LR/SC loop by my reading of the ISA manual

"The dynamic
code executed between the LR and SC instructions can only contain instructions from the base ''I''
instruction set, excluding loads, stores, backward jumps, taken backward branches, JALR, FENCE,
and SYSTEM instructions. If the ''C'' extension is supported, then compressed forms of the
aforementioned ''I'' instructions are also permitted."

topperc commented on llvm/test/CodeGen/RISCV/atomic-rmw.ll:
```diff
@@ -28602,6 +34891,21 @@ define i32 @atomicrmw_max_i32_monotonic(ptr %a, i32 %b) nounwind {
+; RV64I-ZALRSC-NEXT:  .LBB165_1: # =>This Inner Loop Header: Depth=1
+; RV64I-ZALRSC-NEXT:    lr.w a2, (a0)
+; RV64I-ZALRSC-NEXT:    mv a3, a2
+; RV64I-ZALRSC-NEXT:    bge a3, a1, .LBB165_3
```
The upper 32 bits of a1 are garbage.

topperc commented on llvm/test/CodeGen/RISCV/atomic-rmw.ll:
```diff
@@ -29572,6 +36161,22 @@ define i32 @atomicrmw_umax_i32_monotonic(ptr %a, i32 %b) nounwind {
+; RV64I-ZALRSC-NEXT:    lr.w a2, (a0)
+; RV64I-ZALRSC-NEXT:    slli a3, a2, 32
+; RV64I-ZALRSC-NEXT:    srli a3, a3, 32
+; RV64I-ZALRSC-NEXT:    bgeu a3, a1, .LBB175_3
```
The upper 32 bits of a1 are garbage

slachowsky replied:
This is unfortunate.

The type legalizer is handling the 32-bit arguments properly for the `cmpxchg` case because TLI.getExtendForAtomicCmpSwapArg() controls the argument promotion in DAGTypeLegalizer::PromoteIntRes_AtomicCmpSwap(), and RISCV TLI overrides this virtual to return SIGN_EXTEND for correct behavior.

No such control was plumbed for DAGTypeLegalizer::PromoteIntRes_Atomic1() used for `atomicrmw <op>` cases, so we get any extended arguments always.

We could add another virtual TLI function to control this more explicitly for anyone else that might want it.

Or some early RISCV DAGCombine to adjust `atomicrmw {min,max,umin,umax}`  arguments and avoid the default legalizer behavior.

Or solve this in the pseudo expansion, but that is not ideal because I believe it would require a second scratch register.

Thoughts?

slachowsky replied:
Reference to the same issue that plagued `cmpxchg` many years ago: 616289ed29225

arichardson commented on clang/lib/Basic/Targets/RISCV.h:
```diff
@@ -225,7 +225,7 @@ class LLVM_LIBRARY_VISIBILITY RISCV64TargetInfo : public RISCVTargetInfo {
     MaxAtomicPromoteWidth = 128;
 
-    if (ISAInfo->hasExtension("a"))
+    if (ISAInfo->hasExtension("a") || ISAInfo->hasExtension("zalrsc"))
```
Shouldn't a imply zalrsc?

slachowsky replied:
done

slachowsky committed:
[LegalizeTypes][RISCV] Control sign-extend for atomicrmw input argument

Similar to the argument sign-extend control provided by
getExtendForAtomicCmpSwapArg for `cmpxchg`, this change adds
getExtendForAtomicRMWArg for `atomicrmw <op>`.

This mechanism is used to correct the RISCV code generation when using atomic
pseudo expansions that perform sub-word comparions inside a LR/SC block, and
expect a correctly sign-extended input argument.  A shorter instruction
sequence is possible for umin/max by leveraging the fact that two sign-extended
arguments still order correctly during an unsigned comparison.

wangpc-pp commented:
> Is there an abridged history of the RISC-V "A" spec anywhere? I feel reasonably out-of-touch with how it has developed (and been split into sub-extensions).

`A` exists for a long time (maybe since the birth of RISC-V?). The problem is that `A` extension puts two categories of atomics (LR/SC and AMO) into one extension. For some MCUs cores, they would prefer to implement LR/SC (or AMO, if exists) only. So the RVI decided to split `A` into `Zalrsc` and `Zaamo` (https://github.com/riscv/riscv-zaamo-zalrsc/blob/main/zaamo-zalrsc.adoc).  Such things have happened several times, like `M` and `Zmmul`, `C` and `Zca`/`Zcf`/`Zcd`/...

I added these two extensions in the MC level but haven't gotten a chance to support them in CodeGen. :-)

slachowsky commented on llvm/lib/Target/RISCV/RISCVISelLowering.cpp:
```diff
@@ -24483,6 +24483,26 @@ ISD::NodeType RISCVTargetLowering::getExtendForAtomicCmpSwapArg() const {
+    return ISD::SIGN_EXTEND;
+  case ISD::ATOMIC_LOAD_UMIN:
+  case ISD::ATOMIC_LOAD_UMAX:
+    return ISD::ZERO_EXTEND;
```
Done in 72838215e6a7397178483c6e3643c61ef3a17df2

topperc commented on clang/lib/Basic/Targets/RISCV.cpp:
```diff
@@ -192,8 +192,11 @@ void RISCVTargetInfo::getTargetDefines(const LangOptions &Opts,
 
-  if (ISAInfo->hasExtension("a")) {
+  // The "a" extension is composed of "zalrsc" and "zaamo"
+  if (ISAInfo->hasExtension("zalrsc") && ISAInfo->hasExtension("zaamo"))
```
I think this should stay "a". I put up https://github.com/llvm/llvm-project/pull/163890 to make Zalrsc+Zaamo imply A.

slachowsky replied:
Done.

topperc commented on llvm/lib/Target/RISCV/RISCVInstrInfoA.td:
```diff
@@ -308,7 +308,65 @@ class PseudoMaskedAMOMinMaxPat<Intrinsic intrin, Pseudo AMOInst>
            timm:$ordering)>;
 
-let Predicates = [HasStdExtA] in {
+let Predicates = [HasStdExtZalrsc, NoStdExtZaamo] in {
```
There is a comment around line 161 above where `HasAtomicLdSt` is used that says `Atomic load/store are available under both +a and +force-atomics.`. This comment needs to be updated to +zalrsc

fpetrogalli reviewed (approved):
LGTM, thank you.

slachowsky commented on clang/lib/Basic/Targets/RISCV.cpp:
```diff
@@ -192,8 +192,11 @@ void RISCVTargetInfo::getTargetDefines(const LangOptions &Opts,
+  if (ISAInfo->hasExtension("a"))
     Builder.defineMacro("__riscv_atomic");
+
+  if (ISAInfo->hasExtension("zalrsc")) {
```
done, this is checked in a separate preproc test called `riscv-atomics.c`

topperc commented:
I think you lost some of your patch?

slachowsky commented:
> I think you lost some of your patch?

It should be restored, bad rebase.

I don't have contributor access, so GitHub says a maintainer needs to approve the "Build and Test <OS>" workflows, I hope you can oblige.

topperc reviewed (approved):
LGTM
